Components: Add VisuallyHidden component

mkaz opened the pull request:
## Description

Adds a new component for rendering ScreenReaderText and the necessary CSS.
Per @youknowriad <a href="https://github.com/WordPress/gutenberg/pull/17997#issuecomment-543584382">comment here</a>


In testing this, I noticed one of the issues also for Storybook is they weren't setup to load the style.scss along with the components. This PR also adds the necessary CSS loaders in the Storybook webpack.config.js so a story can do this to include the CSS.

```
import ScreenReaderText from '../';
import '../style.scss';

export default { title: 'ScreenReaderText', component: ScreenReaderText };

export const _default = () => (
	<>
		<ScreenReaderText>
			This should not show.
		</ScreenReaderText>
		<div>
			This text will always show.
		</div>
	</>
);
```
This is my first new component being added, so not sure all the pieces that need to be properly updated.

## How has this been tested?

- Add `<ScreenReaderText> Test text. </ScreenReaderText>` and confirm it includes a div and proper style rules.

- Run Storybook: `npm run designs-system:dev` confirm ScreenReaderText does/doesn't show

## Types of changes
- Adds new component `ScreenReaderText`
- Updates Storybook to support loading .scss

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

mkaz committed:
Add ScreenReaderText component

mkaz committed:
Add new component readme to manifest

gziolo commented:
It's often named `VisuallyHidden` in UI components libraries. See:
- https://reakit.io/docs/visually-hidden/ (@diegohaz)
- https://reacttraining.com/reach-ui/visually-hidden

gziolo commented on packages/components/src/screen-reader-text/README.md:
```diff
@@ -0,0 +1,9 @@
+# ScreenReaderText
+
+A component used to render text intended for a screen reader, not visible elsewhere.
```
You can reference WordPress docs here, maybe also copy some parts of that document:
https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/

gziolo commented on packages/components/src/screen-reader-text/README.md:
````diff
@@ -0,0 +1,9 @@
+### Usage
+
+```jsx
+<ScreenReaderText> Show text for screenreader. </ScreenReaderText>
````
In other places we add full examples:

```jsx
import { Dashicon } from '@wordpress/components';

const MyDashicon = () => (
	<div>
		<Dashicon icon="admin-home" />
		<Dashicon icon="products" />
		<Dashicon icon="wordpress" />
	</div>
);
```

gziolo commented on packages/components/src/screen-reader-text/index.js:
```diff
@@ -0,0 +1,10 @@
+
+function ScreenReaderText( props ) {
+	return (
+		<div className="screen-reader-text">
```
You often will want to use a different tag name as well.

mkaz replied:
Instead of a div? I'm open to suggestions, should it b a prop?

youknowriad replied:
The className here doesn't follow our guidelines. I guess it was done this way to match the existing className. I think it's fine to change it for consistency.

youknowriad replied:
In another component we used `isInline` prop to switch between `span` and `div` to ensure HTML stays valid in all situations. I wouldn't mind starting with div and adding that prop if we find a place where we need something inline.

gziolo replied:
The thing is that WP core uses `screen-reader-text`.

See also https://github.com/WordPress/gutenberg/pull/18022#issuecomment-543974141 about `VisualHidden` proposal.

mkaz replied:
One of the reasons for using `screen-reader-text` className was to introduce the CSS into the components built-style CSS. There are several components already using that className relying on core's definition of it.

This was discovered when creating the stories which only include the build-style CSS.

youknowriad replied:
I think we should update all the usage of this class in our codebase to use this component instead, that way it's more portable, the className becomes an implementation detail. It could be done part of this PR or separately.

mkaz replied:
ok, I switched the className to `components-visualyl-hidden` 

I'll switch the other components over in a separate PR. I like keeping each PR as focused as possible and thus one adding a component doesn't really change the other behavior. Just enables the ability to fix them.

gziolo replied:
I think we have `label`, `p` and `leged` with the `screen-reader-text` class name as well.

We can iterate if you prefer to tackle it separately. See my related comment: https://github.com/WordPress/gutenberg/pull/18022#discussion_r337530787.

youknowriad replied:
Are all these solved by nesting VisuallyHidden inline inside the tags?

youknowriad replied:
I think we should decide now whether we want `isInline` prop or a render prop. I feel they address the same issue and I don't like API duplication :)

gziolo replied:
I don't think it solves the issue because the wrapping `label` or `legend` can have additional visual styles applied which you want to remove with `VisuallyHidden`. We would have to double-check that though. I'm making the case based on my assumptions :)

mkaz replied:
You can look at `BaseControl` for an example usage. That solution inserts `<label className='screen-reader-text'>...</label>` 
This makes me think we would want to use the `as` prop. So the implementation would be `<VisuallyHidden as="label">...</VisuallyHidden>`

Also looks like a bug that if you pass in `hideLabelFromVision` prop but do not pass an `id` than the `BaseControl.VisualLabel` will be used, so would not include the class.

mkaz committed:
Remove CSS style loading within stories

mkaz committed:
Switch component name to VisuallyHidden

- Rename directory and includes
- Update README usage
- Update Storybook usage

mkaz committed:
Switch classname to components-visually-hidden

mkaz committed:
Lint: newline

gziolo commented on packages/components/src/visually-hidden/style.scss:
```diff
@@ -0,0 +1,13 @@
+.components-visually-hidden {
```
This is missing:
```css
.screen-reader-text:focus {
  background-color: #eee;
  clip: auto !important;
  clip-path: none;
  color: #444;
  display: block;
  font-size: 1em;
  height: auto;
  left: 5px;
  line-height: normal;
  padding: 15px 23px 14px;
  text-decoration: none;
  top: 5px;
  width: auto;
  z-index: 100000; /* Above WP toolbar. */
}
```

See https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/

mkaz replied:
oh yeh, good catch

gziolo commented on packages/components/src/visually-hidden/index.js:
```diff
@@ -0,0 +1,14 @@
+	children,
+	isInline = false,
+} ) {
+	const TagName = isInline ? 'span' : 'div';
```
Reakit uses an interesting pattern, they have `as` prop which you can pass to change the tag name, or even use a component. See: https://reakit.io/docs/composition/#as-prop

```js
<VisuallyHidden as="span">
	This should not show.
</VisuallyHidden>
```

They also offer an alternative way to use render props. See: https://reakit.io/docs/composition/#render-props

```js
< VisuallyHidden>
      { (  { children } ) => (
        <span {...props}>{ children }</span> 
      ) }
</VisuallyHidden>
```

@diegohaz, how much work would it be to replicate that here?

youknowriad replied:
`as` seem to "web" focused for me especially if we want the API of these components to be DOM agnostic. and the render props are great but I'd like to know if there's a potential use-case that the current API doesn't solve?

gziolo replied:
In Reakit `as` can be a string or Component, the native version:

```js
<VisuallyHidden as={ View }>
    This should not show.
</VisuallyHidden>
```

youknowriad replied:
mmm interesting

diegohaz replied:
It's simple to implement:
```jsx
function renderAsRenderProps({ as: T = "div", ...props }) {
  if (typeof props.children === "function") {
    return props.children(props);
  }
  return <T {...props} />;
}

function Component({ as = "span", ...props }) {
  return renderAsRenderProps({ as, ...props });
}
```

Using children as a function is better than the `as` prop to compose multiple components:
```jsx
<Component>
  {props => <div><Button {...props} /></div>}
</Component>
```

diegohaz replied:
The `as` prop is quite popular nowadays. The first library I saw using it was [Semantic UI](https://react.semantic-ui.com/#augmentation). Not sure if there are other libraries combining it with render props though. But I think it's a simple and powerful combination.

mkaz replied:
Thanks @diegohaz - I updated the component to use the "as" method as described, works well.

I defined the `renderAsRenderProps` function within the VisuallyHidden component, not exported. It  seems useful that we may want to use it elsewhere. If so, where would I move it to?

gziolo reviewed (changes requested):
The missing CSS styles for the focused state is essential to make it work.

mkaz committed:
Add focus style

mkaz committed:
Switch to 'as' for specifying tag

youknowriad commented on packages/components/src/visually-hidden/index.js:
```diff
@@ -0,0 +1,24 @@
+}
+
+// render "as" to be either string (tag name) or component (fn)
+function renderAsRenderProps( { as: T = 'div', ...props } ) {
```
I like that pattern a lot and this can probably be extracted in a utils folder in the components package. That said a `<Component as="div" />` doesn't mean anything for mobile since there's no "div" on mobile and there's the expectation that the same props should be valid in all platform.

Not a blocker since mobile can decide to just ignore the "as"  prop but I thought I'd raise for awareness.

mkaz replied:
I moved it to a `utils.js` file.

The div is just the default value, if used on Mobile they can pass in something different, or not use :-)

youknowriad replied:
The idea is that mobile and web use the exact same code in the `edit` function of blocks... So there's no mobile do that and web do this.

mkaz replied:
@youknowriad ah, I see.  Sounds like its not a blocker though, so should be fine.
Is the utils.js extraction fine?

mkaz replied:
Actually @gziolo has the change requested for the CSS. Greg, when you get the chance, can you do a quick review again. I think everything has been addressed. Thanks! :bowing_man:

mkaz committed:
Move renderAsRenderProps to utils.js

diegohaz commented:
It looks great! I'd say only that the `renderAsRenderProps` util could be kept inside the component folder for now. There is a few things we should define before making it available for other components (naming, React Native version etc.), and that could be done in another PR. If this PR is merged and people start using the util in other components, it'll be harder to change later.

mkaz committed:
Move utils to inside component folder

Waiting to refine the utils usage a little better before
making it look available for other components to use.

mkaz commented:
@diegohaz Thanks. I moved the utils file, inside VisuallyHidden component folder for now. We can adjust as things get defined.

gziolo commented on packages/components/src/visually-hidden/utils.js:
```diff
@@ -0,0 +1,22 @@
+ * @param {string|Component} as what tag/component to render
+ * @return {Component} rendered component
+ */
+function renderAsRenderProps( { as: T = 'div', ...props } ) {
```
Can we avoid using shortcuts, please. It makes it hard to follow for those not familiar with the pattern:
```suggestion
function renderAsRenderProps( { as: Component = 'div', ...props } ) {
```

gziolo commented on packages/components/src/visually-hidden/utils.js:
```diff
@@ -0,0 +1,22 @@
+ *
+ * See VisuallyHidden hidden for example.
+ *
+ * @param {string|Component} as what tag/component to render
```
Nit:
```suggestion
 * @param {string|Component} as A tag or component to render.
```

gziolo commented on packages/components/src/visually-hidden/utils.js:
```diff
@@ -0,0 +1,22 @@
+ * See VisuallyHidden hidden for example.
+ *
+ * @param {string|Component} as what tag/component to render
+ * @return {Component} rendered component
```
```suggestion
 * @return {Component} The rendered component.
```

gziolo commented on packages/components/src/visually-hidden/index.js:
```diff
@@ -0,0 +1,22 @@
+
```
This empty space is not necessary. Why is ESLint not complaining? :) We really need Prettier 😍

mkaz replied:
Funny because the space should be below, newline at the end.

gziolo commented on packages/components/src/visually-hidden/style.scss:
```diff
@@ -0,0 +1,30 @@
+}
+
+.components-visually-hidden:focus {
+	background-color: #eee;
```
Can this be converted to one of the variables used in SCSS files?

mkaz replied:
This raises another issue, it doesn't look like the SCSS variables are defined within Components package, or if they are I've missed where. It seems they rely on Block Editor styling, which can be used, but not sure where this separation should be. So this might need to become a separate issue for how to define and use SCSS with Components.

youknowriad replied:
That's an already solved problem. variables are now defined in the `base-styles` package and these are made available to all packages using SASS including the @wordpress/components package.

mkaz replied:
:+1: Great. So sounds like mostly a matter of documentation and examples for the case that someone wants to use the components package, show they would define their own or use base-styles package.

gziolo commented on packages/components/src/visually-hidden/style.scss:
```diff
@@ -0,0 +1,30 @@
+	text-decoration: none;
+	top: 5px;
+	width: auto;
+	z-index: 100000;
```
In the context of Gutenberg, we probably should use a special helper which processes all `z-indexes`:

https://github.com/WordPress/gutenberg/blob/7515397fef469e8cbdaf68162d6d6e1f2a5f9aa1/packages/base-styles/_z-index.scss#L127-L133

gziolo reviewed (approved):
I listed a few nits to address, those aren't blockers if you plan to work on a follow-up PR. Your call @mkaz.

Thank you for addressing all the feedback, it was a journey to get us to this stage but I think it's worth it. Props to @diegohaz for help with the `renderAsRenderProps` integration 💯

mkaz committed:
Apply suggestions from code review

Co-Authored-By: Grzegorz (Greg) Ziółkowski <grzegorz@gziolo.pl>

mkaz committed:
Lint: Move newline

mkaz committed:
Fix variable name

mkaz committed:
Use variable for stylesheet
